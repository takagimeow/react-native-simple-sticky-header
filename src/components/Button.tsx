import React from 'react';
import { Text, TouchableOpacity, View } from 'react-native';

export function Button({ children, active = false }: { children: string; active?: boolean }) {
  return (
    <TouchableOpacity>
      <View
        style={{
          borderRadius: 25,
          backgroundColor: active ? '#333333' : '#ffffff',
          paddingVertical: 12,
          paddingHorizontal: 16,
          marginHorizontal: 4,
          width: 116,
        }}
      >
        <Text
          style={{
            color: active ? '#ffffff' : '#333333',
            textAlignVertical: 'center',
            textAlign: 'center',
          }}
        >
          {children}
        </Text>
      </View>
    </TouchableOpacity>
  );
}
