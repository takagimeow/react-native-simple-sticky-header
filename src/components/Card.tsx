import React from 'react';
import { Dimensions, StyleProp, Text, View, ViewStyle } from 'react-native';

export interface ICard {
  title: string;
  description: string;
  score: number;
  style?: StyleProp<ViewStyle>;
}

export function Card({ title, description, score, style }: ICard) {
  return (
    <View
      style={[
        {
          width: Dimensions.get('screen').width,
          backgroundColor: '#fffffff',
          paddingTop: 12,
          paddingBottom: 16,
          paddingHorizontal: 14,
        },
        style,
      ]}
    >
      <View
        style={{
          width: '100%',
          height: 156,
          backgroundColor: '#f0f0f0',
          marginBottom: 12,
        }}
      ></View>
      <View
        style={{
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'space-between',
        }}
      >
        <Text
          style={{
            fontSize: 14,
          }}
        >
          {title.length > 25 ? `${title.slice(0, 25)}...` : title}
        </Text>
        <View
          style={{
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            borderRadius: 25,
            paddingVertical: 6,
            paddingHorizontal: 4,
            backgroundColor: '#f0f0f0',
          }}
        >
          <Text
            style={{
              fontSize: 12,
            }}
          >
            {score}
          </Text>
        </View>
      </View>
      <View
        style={{
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'space-between',
        }}
      >
        <Text
          style={{
            fontSize: 12,
          }}
        >
          {description}
        </Text>
      </View>
    </View>
  );
}
