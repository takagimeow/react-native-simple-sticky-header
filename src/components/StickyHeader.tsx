import React from 'react';
import { Animated } from 'react-native';

export interface IStickyHeader {
  children: React.ReactElement;
  style?: {
    height?: number | string;
  };
}

export function StickyHeader({ children, style }: IStickyHeader) {
  return (
    <Animated.View
      style={{
        ...style,
        width: '100%',
        alignItems: 'center',
        justifyContent: 'center',
      }}
    >
      {children}
    </Animated.View>
  );
}
