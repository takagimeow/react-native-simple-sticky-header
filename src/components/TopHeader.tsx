import React from 'react';
import { Animated, LayoutChangeEvent } from 'react-native';

export interface ITopHeader {
  children: React.ReactElement;
  opacity?: number | Animated.Value | Animated.AnimatedInterpolation | undefined;
  style?: {
    height?: number | string;
  };
  onLayout?: ((event: LayoutChangeEvent) => void) | undefined;
}

export function TopHeader({ children, opacity, style, onLayout }: ITopHeader) {
  return (
    <Animated.View onLayout={onLayout}>
      <Animated.View
        style={{
          ...style,
          width: '100%',
          alignItems: 'center',
          justifyContent: 'center',
          opacity,
        }}
      >
        {children}
      </Animated.View>
    </Animated.View>
  );
}
