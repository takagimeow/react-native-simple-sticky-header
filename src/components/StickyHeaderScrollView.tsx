import React, { useState, useRef, useEffect } from 'react';
import {
  StatusBar,
  SafeAreaView,
  Animated,
  Dimensions,
  RefreshControl,
  Platform,
  View,
} from 'react-native';
import { getStatusBarHeight } from 'react-native-status-bar-height';
import { StickyHeader } from './StickyHeader';
import { TopHeader } from './TopHeader';

export interface IStickyHeaderScrollView {
  children: React.ReactElement;
  top: (ref: React.LegacyRef<View>) => React.ReactElement;
  bottom: (ref: React.LegacyRef<View>) => React.ReactElement;
  statusBarBackground?: string;
  headerBackground?: string;
  scrollViewBackground?: string;
  refreshing?: boolean;
  onRefresh?: () => void;
}

export function StickyHeaderScrollView({
  children,
  top,
  bottom,
  statusBarBackground = '#C4C4C4',
  headerBackground = '#ffffff',
  scrollViewBackground = '#ffffff',
  refreshing = false,
  onRefresh,
}: IStickyHeaderScrollView) {
  const [topHeight, setTopHeight] = useState(0);
  const [bottomHeight, setBottomHeight] = useState(0);
  const headerComponentRef = useRef(null);
  const topComponentRef = useRef(null);
  const bottomComponentRef = useRef(null);
  const scrollViewRef = useRef(new Animated.Value(0));
  const scrollY = useRef(new Animated.Value(0)).current;
  const translateYRef = useRef(new Animated.Value(0));
  const opacityRef = useRef(new Animated.Value(0));

  const [statusBarHeight] = useState(getStatusBarHeight());
  const [headerHeight, setHeaderHeight] = useState(topHeight + bottomHeight + statusBarHeight);
  const [paddingTop, setPaddingTop] = useState(0);
  const lastTimeOffsetRef = useRef(0);
  const animatingRef = useRef(false);
  useEffect(() => {
    const newHeaderHeight =
      Platform.OS === 'android'
        ? topHeight + bottomHeight + statusBarHeight
        : topHeight + bottomHeight;
    const paddingTop =
      Platform.OS === 'android'
        ? topHeight !== bottomHeight
          ? newHeaderHeight + Math.abs(topHeight - bottomHeight)
          : newHeaderHeight
        : 0;
    setHeaderHeight(newHeaderHeight);
    setPaddingTop(paddingTop);
  }, [topHeight, headerHeight, scrollViewRef]);

  useEffect(() => {
    scrollY.addListener(event => {
      if (animatingRef.current) {
        return;
      }
      console.log('event.value: ', event.value);
      console.log('lastTimeOffsetRef.current: ', lastTimeOffsetRef.current);
      console.log('\n\n\n');
      if (
        event.value <= 0 ||
        (event.value <= lastTimeOffsetRef.current && event.value < Dimensions.get('screen').height)
      ) {
        // 点灯
        Animated.parallel([
          Animated.timing(translateYRef.current, {
            toValue: 0,
            duration: 160,
            useNativeDriver: true,
          }),
          Animated.timing(opacityRef.current, {
            toValue: 1,
            duration: 160,
            useNativeDriver: true,
          }),
        ]).start(({ finished }) => {
          animatingRef.current = !finished;
        });
      } else {
        // 消灯
        Animated.parallel([
          Animated.timing(translateYRef.current, {
            toValue: -topHeight,
            duration: 160,
            useNativeDriver: true,
          }),
          Animated.timing(opacityRef.current, {
            toValue: 0,
            duration: 160,
            useNativeDriver: true,
          }),
        ]).start(({ finished }) => {
          animatingRef.current = !finished;
        });
      }
      lastTimeOffsetRef.current = event.value;
    });

    return () => {
      scrollY.removeAllListeners();
    };
  }, [topHeight, headerHeight, headerComponentRef]);

  return (
    <>
      <SafeAreaView
        style={{
          flex: 1,
          backgroundColor: scrollViewBackground,
        }}
      >
        <Animated.ScrollView
          ref={scrollViewRef}
          refreshControl={
            onRefresh ? (
              <RefreshControl
                progressViewOffset={headerHeight}
                refreshing={refreshing}
                onRefresh={onRefresh}
              />
            ) : undefined
          }
          onScroll={Animated.event(
            [
              {
                nativeEvent: {
                  contentOffset: { y: scrollY },
                },
              },
            ],
            { useNativeDriver: true },
          )}
          scrollEventThrottle={16}
          contentInset={{
            top:
              Platform.OS === 'ios'
                ? topHeight !== bottomHeight
                  ? headerHeight + Math.abs(topHeight - bottomHeight)
                  : headerHeight
                : 0,
          }}
          contentOffset={{
            x: 0,
            y:
              Platform.OS === 'ios'
                ? topHeight !== bottomHeight
                  ? -(headerHeight + Math.abs(topHeight - bottomHeight))
                  : -headerHeight
                : 0,
          }}
          contentContainerStyle={{
            paddingTop,
          }}
          automaticallyAdjustContentInsets={false}
          contentInsetAdjustmentBehavior="automatic"
          showsVerticalScrollIndicator={false}
          showsHorizontalScrollIndicator={false}
        >
          <Animated.View
            style={{
              backgroundColor: 'transparent',
            }}
          >
            {children}
          </Animated.View>
        </Animated.ScrollView>
      </SafeAreaView>
      <Animated.View
        ref={headerComponentRef}
        style={[
          {
            position: 'absolute',
            top: statusBarHeight,
            width: Dimensions.get('screen').width,
            backgroundColor: headerBackground,
            transform: [
              {
                translateY: translateYRef.current,
              },
            ],
          },
        ]}
      >
        <TopHeader
          opacity={opacityRef.current}
          onLayout={event => {
            const height = event.nativeEvent.layout.height;
            setTopHeight(height);
          }}
        >
          <View>{top(topComponentRef)}</View>
        </TopHeader>
        <StickyHeader
          onLayout={event => {
            const height = event.nativeEvent.layout.height;
            setBottomHeight(height);
          }}
        >
          {bottom(bottomComponentRef)}
        </StickyHeader>
      </Animated.View>
      <StatusBar backgroundColor={statusBarBackground} translucent={true} />
    </>
  );
}
