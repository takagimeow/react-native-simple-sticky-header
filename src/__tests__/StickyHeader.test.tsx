/* eslint-disable @typescript-eslint/unbound-method */
import React from 'react';
import 'react-native';
import { render } from '@testing-library/react-native';

import { StickyHeader } from '../components/StickyHeader';
import { Text, TouchableOpacity, View } from 'react-native';

describe('StickyHeader', () => {
  it('Snapshot Test', () => {
    const { toJSON } = render(
      <StickyHeader onLayout={jest.fn()}>
        <View
          style={{
            display: 'flex',
            flexDirection: 'row',
            justifyContent: 'space-around',
            width: '100%',
            alignItems: 'center',
            height: 64,
          }}
        >
          <View
            style={{
              marginTop: 2,
              width: 16,
              height: 16,
            }}
          ></View>
          <View>
            <TouchableOpacity>
              <Text
                style={{
                  fontSize: 16,
                }}
              >
                今すぐ・東京都千代田区1丁目1番地
              </Text>
            </TouchableOpacity>
          </View>
          <View
            style={{
              marginTop: 2,
            }}
          >
            <TouchableOpacity>
              <Text
                style={{
                  fontSize: 16,
                }}
              >
                👾
              </Text>
            </TouchableOpacity>
          </View>
        </View>
      </StickyHeader>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
