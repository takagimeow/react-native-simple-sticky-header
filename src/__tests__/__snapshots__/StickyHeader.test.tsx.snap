// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickyHeader Snapshot Test 1`] = `
<View
  collapsable={false}
  nativeID="animatedComponent"
  onLayout={[MockFunction]}
  style={Object {}}
>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "height": 64,
          "justifyContent": "space-around",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "height": 16,
            "marginTop": 2,
            "width": 16,
          }
        }
      />
      <View>
        <View
          accessible={true}
          collapsable={false}
          focusable={false}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
              }
            }
          >
            今すぐ・東京都千代田区1丁目1番地
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "marginTop": 2,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={false}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
              }
            }
          >
            👾
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
